### PR TITLE
[SPARK-5979][SPARK-6031][SPARK-6032][SPARK-6047] Refactoring for --packages -> Move to SparkSubmitDriverBootstrapper

### PR DESCRIPTION
**Pull request by brkyvz**
This PR is an umbrella PR for 3 JIRAs. Here're the explanations:
- SPARK-5979: All dependencies with the groupId `org.apache.spark` passed through `--packages`, were being excluded from the dependency tree on the assumption that they would be in the assembly jar. This is not the case, therefore the exclusion rules had to be defined more explicitly.
- SPARK-6031: When using pyspark or running a python program through spark-submit, py4j was not picking up the dynamically loaded jars on the driver. Moving the code for `--packages` to SparkSubmitDriverBootstrapper solves this. However, this issue still remains for `--jars`.
- SPARK-6032: Ivy prints a whole lot of logs while retrieving dependencies. These were printed to `System.out`. Moved the logging to `System.err`.

@tdas Would you care to try this? I think it should solve your problem

### CODE REVIEW

**Comment by tdas** [comment]
@pwendell Please take a look at this. I think you reviewed the original PR of this feature.

**Comment by pwendell** [comment]
LGTM

**Comment by brkyvz** [comment]
@tdas added a hack to include the jars on --driver-extra-classpath. Can you try your test now?

**Comment by tdas** [comment]
I tested. Still not working. I enabled verbose logging on spark-submit and saw this 

```
[tdas @ Zion spark2] bin/spark-submit --verbose --master local[4] --repositories https://repository.apache.org/content/repositories/orgapachespark-1069/ --packages org.apache.spark:spark-streaming-kafka_2.10:1.3.0 examples/src/main/python/streaming/kafka_wordcount.py localhost:2181 test
Using properties file: null
Parsed arguments:
  master                  local[4]
  deployMode              null
  executorMemory          null
  executorCores           null
  totalExecutorCores      null
  propertiesFile          null
  driverMemory            null
  driverCores             null
  driverExtraClassPath    null
  driverExtraLibraryPath  null
  driverExtraJavaOptions  null
  supervise               false
  queue                   null
  numExecutors            null
  files                   null
  pyFiles                 null
  archives                null
  mainClass               null
  primaryResource         file:/Users/tdas/Projects/Spark/spark2/examples/src/main/python/streaming/kafka_wordcount.py
  name                    kafka_wordcount.py
  childArgs               [localhost:2181 test]
  jars                    null
  packages                org.apache.spark:spark-streaming-kafka_2.10:1.3.0
  repositories            https://repository.apache.org/content/repositories/orgapachespark-1069/
  verbose                 true

Spark properties used, including those specified through
 --conf and those from the properties file null:



Ivy Default Cache set to: /Users/tdas/.ivy2/cache
The jars for the packages stored in: /Users/tdas/.ivy2/jars
https://repository.apache.org/content/repositories/orgapachespark-1069/ added as a remote repository with the name: repo-1
:: loading settings :: url = jar:file:/Users/tdas/Projects/Spark/spark2/assembly/target/scala-2.10/spark-assembly-1.3.0-SNAPSHOT-hadoop1.0.4.jar!/org/apache/ivy/core/settings/ivysettings.xml
org.apache.spark#spark-streaming-kafka_2.10 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent;1.0
	confs: [default]
	found org.apache.spark#spark-streaming-kafka_2.10;1.3.0 in repo-1
	found org.apache.kafka#kafka_2.10;0.8.1.1 in list
	found com.yammer.metrics#metrics-core;2.2.0 in list
	found org.slf4j#slf4j-api;1.7.10 in list
	found org.xerial.snappy#snappy-java;1.1.1.6 in list
	found com.101tec#zkclient;0.3 in list
	found log4j#log4j;1.2.17 in list
	found org.spark-project.spark#unused;1.0.0 in list
:: resolution report :: resolve 370ms :: artifacts dl 17ms
	:: modules in use:
	com.101tec#zkclient;0.3 from list in [default]
	com.yammer.metrics#metrics-core;2.2.0 from list in [default]
	log4j#log4j;1.2.17 from list in [default]
	org.apache.kafka#kafka_2.10;0.8.1.1 from list in [default]
	org.apache.spark#spark-streaming-kafka_2.10;1.3.0 from repo-1 in [default]
	org.slf4j#slf4j-api;1.7.10 from list in [default]
	org.spark-project.spark#unused;1.0.0 from list in [default]
	org.xerial.snappy#snappy-java;1.1.1.6 from list in [default]
	---------------------------------------------------------------------
	|                  |            modules            ||   artifacts   |
	|       conf       | number| search|dwnlded|evicted|| number|dwnlded|
	---------------------------------------------------------------------
	|      default     |   8   |   0   |   0   |   0   ||   8   |   0   |
	---------------------------------------------------------------------
:: retrieving :: org.apache.spark#spark-submit-parent
	confs: [default]
	0 artifacts copied, 8 already retrieved (0kB/7ms)
Main class:
org.apache.spark.deploy.PythonRunner
Arguments:
file:/Users/tdas/Projects/Spark/spark2/examples/src/main/python/streaming/kafka_wordcount.py
/Users/tdas/.ivy2/jars/spark-streaming-kafka_2.10.jar,/Users/tdas/.ivy2/jars/kafka_2.10.jar,/Users/tdas/.ivy2/jars/unused.jar,/Users/tdas/.ivy2/jars/metrics-core.jar,/Users/tdas/.ivy2/jars/snappy-java.jar,/Users/tdas/.ivy2/jars/zkclient.jar,/Users/tdas/.ivy2/jars/slf4j-api.jar,/Users/tdas/.ivy2/jars/log4j.jar
localhost:2181
test
System properties:
SPARK_SUBMIT -> true
spark.submit.pyFiles -> /Users/tdas/.ivy2/jars/spark-streaming-kafka_2.10.jar,/Users/tdas/.ivy2/jars/kafka_2.10.jar,/Users/tdas/.ivy2/jars/unused.jar,/Users/tdas/.ivy2/jars/metrics-core.jar,/Users/tdas/.ivy2/jars/snappy-java.jar,/Users/tdas/.ivy2/jars/zkclient.jar,/Users/tdas/.ivy2/jars/slf4j-api.jar,/Users/tdas/.ivy2/jars/log4j.jar
spark.files -> file:/Users/tdas/Projects/Spark/spark2/examples/src/main/python/streaming/kafka_wordcount.py,file:/Users/tdas/.ivy2/jars/spark-streaming-kafka_2.10.jar,file:/Users/tdas/.ivy2/jars/kafka_2.10.jar,file:/Users/tdas/.ivy2/jars/unused.jar,file:/Users/tdas/.ivy2/jars/metrics-core.jar,file:/Users/tdas/.ivy2/jars/snappy-java.jar,file:/Users/tdas/.ivy2/jars/zkclient.jar,file:/Users/tdas/.ivy2/jars/slf4j-api.jar,file:/Users/tdas/.ivy2/jars/log4j.jar
spark.app.name -> kafka_wordcount.py
spark.jars -> file:/Users/tdas/.ivy2/jars/spark-streaming-kafka_2.10.jar,file:/Users/tdas/.ivy2/jars/kafka_2.10.jar,file:/Users/tdas/.ivy2/jars/unused.jar,file:/Users/tdas/.ivy2/jars/metrics-core.jar,file:/Users/tdas/.ivy2/jars/snappy-java.jar,file:/Users/tdas/.ivy2/jars/zkclient.jar,file:/Users/tdas/.ivy2/jars/slf4j-api.jar,file:/Users/tdas/.ivy2/jars/log4j.jar
spark.master -> local[4]
spark.driver.extraClassPath -> /Users/tdas/.ivy2/jars/spark-streaming-kafka_2.10.jar,/Users/tdas/.ivy2/jars/kafka_2.10.jar,/Users/tdas/.ivy2/jars/unused.jar,/Users/tdas/.ivy2/jars/metrics-core.jar,/Users/tdas/.ivy2/jars/snappy-java.jar,/Users/tdas/.ivy2/jars/zkclient.jar,/Users/tdas/.ivy2/jars/slf4j-api.jar,/Users/tdas/.ivy2/jars/log4j.jar
Classpath elements:
/Users/tdas/.ivy2/jars/spark-streaming-kafka_2.10.jar
/Users/tdas/.ivy2/jars/kafka_2.10.jar
/Users/tdas/.ivy2/jars/unused.jar
/Users/tdas/.ivy2/jars/metrics-core.jar
/Users/tdas/.ivy2/jars/snappy-java.jar
/Users/tdas/.ivy2/jars/zkclient.jar
/Users/tdas/.ivy2/jars/slf4j-api.jar
/Users/tdas/.ivy2/jars/log4j.jar

```

So i can see that the relevant jars are being added to the classpath elements but pyspark is still unable to find org.apache.spark.streaming.kafka.KafkaUtils (from /Users/tdas/.ivy2/jars/spark-streaming-kafka_2.10.jar).

Lets debug this tomorrow morning.

**Comment by brkyvz** [comment]
@tdas I think that's returning an error now, because the class in fact doesn't really exist in the jar. Not because it's not in the classpath

**Comment by brkyvz** [comment]
nvm, it should be in spark-streaming-kafka_2.10.jar

**Comment by tdas** [comment]
No I verified the class does exist in jar

**Comment by brkyvz** [comment]
@tdas @pwendell  @andrewor14
This is ready for code review. Moved the resolve method to DriverBootstrapper. In case the DriverBootstrapper is not called, the resolution takes place inside SparkSubmit like before. This should work (tested it with TD's example + my own package examples)

**Comment by tdas** [comment]
Jenkins, test this again.

**Comment by brkyvz** [comment]
It might not be a flaky test. I might have broken some Yarn feature. I'm
going to check once I get home.

**Comment by tdas** [comment]
Ohh... okay.

**Comment by brkyvz** [comment]
@tdas The latest commit fixed the issue, feel free to test

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 1]
```diff
@@ -251,24 +251,17 @@ object SparkSubmit {
...
-      }
-      if (args.isPython) {
-        if (args.pyFiles == null || args.pyFiles.trim.isEmpty) {
-          args.pyFiles = resolvedMavenCoordinates
-        } else {
-          args.pyFiles += s",$resolvedMavenCoordinates"
-        }
+    if (args.packagesResolved != null) {
+      args.jars = mergeFileLists(args.jars, args.packagesResolved)
+      if (args.isPython) args.pyFiles = mergeFileLists(args.pyFiles, args.packagesResolved)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]:
```
if (args.isPython) {
  args.pyFiles = ...
}
```

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 2]
```diff
@@ -764,10 +785,13 @@ private[spark] object SparkSubmitUtils {
       coordinates: String,
       remoteRepos: Option[String],
       ivyPath: Option[String],
-      isTest: Boolean = false): String = {
+      isTest: Boolean = false): Seq[String] = {
     if (coordinates == null || coordinates.trim.isEmpty) {
-      ""
+      Seq("")
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Did you mean to include a single element sequence? Or did you mean to do `Seq.empty`

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 3]
```diff
@@ -764,10 +785,13 @@ private[spark] object SparkSubmitUtils {
       coordinates: String,
       remoteRepos: Option[String],
       ivyPath: Option[String],
-      isTest: Boolean = false): String = {
+      isTest: Boolean = false): Seq[String] = {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: you'll need to update the javadoc to reflect this change

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 4]
```diff
@@ -725,13 +717,13 @@ private[spark] object SparkSubmitUtils {
    */
   private[spark] def resolveDependencyPaths(
       artifacts: Array[AnyRef],
-      cacheDirectory: File): String = {
+      cacheDirectory: File): Seq[String] = {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: same here, update java doc

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala [hunk 5]
```diff
@@ -251,24 +251,17 @@ object SparkSubmit {
...
+      if (args.isPython) args.pyFiles = mergeFileLists(args.pyFiles, args.packagesResolved)
+    } else {
+      // This part needs to be added in case SparkSubmitDriverBootstrapper is not called
+      val resolvedMavenCoordinates =
+        SparkSubmitUtils.resolveMavenCoordinates(
+          args.packages, Option(args.repositories), Option(args.ivyRepoPath)).mkString(",")
+      if (resolvedMavenCoordinates.trim.length > 0) {
+        args.jars = mergeFileLists(args.jars, resolvedMavenCoordinates)
+        if (args.isPython) args.pyFiles = mergeFileLists(args.pyFiles, resolvedMavenCoordinates)
       }
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala]: Looks like there is an opportunity for abstracting the two cases here:
```
val packagesResolved =
  if (args.packagesResolved != null) {
    args.packagesResolved
  } else {
    SparkSubmitUtils.resolveMavenCoordinates(...).mkString(",")
  }
if (packagesResolved.nonEmpty) {
  args.jars = mergeFileLists(args.jars, packagesResolved)
  if (args.isPython) {
    args.pyFiles = mergeFileLists(args.pyFiles, packagesResolved)
  }
}
```

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitDriverBootstrapper.scala [hunk 6]
```diff
@@ -82,13 +85,25 @@ private[spark] object SparkSubmitDriverBootstrapper {
...
-    val newClasspath =
+    var newClasspath =
       if (submitClasspath.isDefined) {
         classpath
       } else {
         classpath + confClasspath.map(sys.props("path.separator") + _).getOrElse("")
       }
 
+    // Resolve maven dependencies if there are any and add classpath to jars. Add them to py-files
+    // too for packages that include Python code
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitDriverBootstrapper.scala]: where is the code that adds it to pyFiles?

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitDriverBootstrapper.scala [hunk 7]
```diff
@@ -82,13 +85,25 @@ private[spark] object SparkSubmitDriverBootstrapper {
...
+    val resolvedMavenCoordinates =
+      SparkSubmitUtils.resolveMavenCoordinates(
+        submitPackages, submitRepositories, confIvyRepo)
+    if (resolvedMavenCoordinates.head.length > 0) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitDriverBootstrapper.scala]: should this be `resolvedMavenCoordinates.nonEmpty`

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala [hunk 8]
```diff
@@ -395,6 +396,11 @@ private[spark] class SparkSubmitArguments(args: Seq[String], env: Map[String, St
         packages = value
         parse(tail)
 
+      // Internal flag to receive the resolved maven jars and add to --jars
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala]: maybe add that this is only passed in from the bootstrapper

**Comment by tdas** [comment]
@brkyvz You probably need to update the title and JIRA, because this change is quite a bit of refactoring rather than just updating of the exclusion rules.

**Comment by andrewor14** [comment]
@brkyvz Spark submit changes LGTM. I am not familiar enough with the maven stuff to comment there.

**Comment by brkyvz** [comment]
This passed locally. What the...

**Comment by andrewor14** [comment]
retest this please

**Comment by brkyvz** [comment]
Flaky test this time... @tdas, can you have this retested please?

**Comment by srowen** [comment]
Jenkins, retest this please

**Comment by brkyvz** [comment]
@srowen Thank you!

**Comment by tdas** [comment]
@brkyvz I think you need to address a couple of more JIRAs in this PR. 4 aint enough ;)

File: core/src/main/scala/org/apache/spark/deploy/SparkSubmitDriverBootstrapper.scala [hunk 9]
```diff
@@ -82,13 +85,25 @@ private[spark] object SparkSubmitDriverBootstrapper {
...
+    // Resolve maven dependencies if there are any and add them to classpath. Also send them
+    // to SparkSubmit so that they can be shipped to executors.
+    val resolvedMavenCoordinates =
+      SparkSubmitUtils.resolveMavenCoordinates(
+        submitPackages, submitRepositories, confIvyRepo)
+    if (resolvedMavenCoordinates.nonEmpty) {
+      newClasspath += sys.props("path.separator") + 
+        resolvedMavenCoordinates.mkString(sys.props("path.separator"))
+      submitArgs = 
+        Array("--packages-resolved", resolvedMavenCoordinates.mkString(",")) ++ submitArgs
```
Comment (pwendell) [inline core/src/main/scala/org/apache/spark/deploy/SparkSubmitDriverBootstrapper.scala]: Can we thread this through using an environment variable `_PACKAGES_RESOLVED`? Having this as an extra flag forces you to make `args` here mutable, which is sort of strange.

**Comment by pwendell** [comment]
@brkyvz let's close this issue for now and keep it in our back pocket. We can use it if we decide to put this in the 1.3 branch down the line.

**Pull request closed without merging**